File: 1_two_sum.py
# Given an array of integers nums and an integer target,
# return indices of the two numbers such that they 
# add up to target.

# You may assume that each input would have exactly one solution, 
# and you may not use the same element twice.

# You can return the answer in any order


nums, target = [2, 7, 11, 15], 9
nums_2, target_2 = [1, 2, 3, 4], 8

def two_sum(nums: list, target) -> list:
    seen = {}
    N = len(nums)
    for i in range(N):
        difference = target - nums[i]
        if difference in seen:
            return [seen[difference], i]
        else:
            seen[nums[i]] = i
    return []
     
print(two_sum(nums, target))
print(two_sum(nums_2, target_2))
     

File: Javascript/1_two_sum.js
const twoSum = function (nums, target) {
  let seens = [];
  for (let i = 0; i < nums.length; i++) {
    console.log(`iteration? ${i}`);
    let difference = target - nums[i];
    console.log(`diff ${difference}`, difference);
    if (seens[difference]) {
      return [seens[difference], i];
    } else {
      seens.push(i);
    }
  }
  return [];
};

const numbers = [2, 7, 11, 15];
const target = 9;

const numbers2 = [8, 4, 1, 19];
const target2 = 10;
console.log(twoSum(numbers, target));
console.log(twoSum(numbers2, target2));
